compress.go

```diff
@@ -0,0 +1,217 @@
+package mysql
+
+import (
+	"bytes"
+	"compress/zlib"
+	"io"
+)
+
+const (
+	minCompressLength = 50
+)
+
+type packetReader interface {
+	readNext(need int) ([]byte, error)
+}
+
+type compressedReader struct {
+	buf      packetReader
+	bytesBuf []byte
+	mc       *mysqlConn
+}
+
+type compressedWriter struct {
+	connWriter io.Writer
+	mc         *mysqlConn
+}
+
+func NewCompressedReader(buf packetReader, mc *mysqlConn) *compressedReader {
+	return &compressedReader{
+		buf:      buf,
+		bytesBuf: make([]byte, 0),
+		mc:       mc,
+	}
+}
+
+func NewCompressedWriter(connWriter io.Writer, mc *mysqlConn) *compressedWriter {
+	return &compressedWriter{
+		connWriter: connWriter,
+		mc:         mc,
+	}
+}
+
+func (cr *compressedReader) readNext(need int) ([]byte, error) {
+	for len(cr.bytesBuf) < need {
+		err := cr.uncompressPacket()
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	data := make([]byte, need)
+
+	copy(data, cr.bytesBuf[:len(data)])
+
+	cr.bytesBuf = cr.bytesBuf[len(data):]
+
+	return data, nil
+}
+
+func (cr *compressedReader) uncompressPacket() error {
+	header, err := cr.buf.readNext(7) // size of compressed header
+
+	if err != nil {
+		return err
+	}
+
+	// compressed header structure
+	comprLength := int(uint32(header[0]) | uint32(header[1])<<8 | uint32(header[2])<<16)
+	uncompressedLength := int(uint32(header[4]) | uint32(header[5])<<8 | uint32(header[6])<<16)
+	compressionSequence := uint8(header[3])
+
+	if compressionSequence != cr.mc.compressionSequence {
+		return ErrPktSync
+	}
+
+	cr.mc.compressionSequence++
+
+	comprData, err := cr.buf.readNext(comprLength)
+	if err != nil {
+		return err
+	}
+
+	// if payload is uncompressed, its length will be specified as zero, and its
+	// true length is contained in comprLength
+	if uncompressedLength == 0 {
+		cr.bytesBuf = append(cr.bytesBuf, comprData...)
+		return nil
+	}
+
+	// write comprData to a bytes.buffer, then read it using zlib into data
+	var b bytes.Buffer
+	b.Write(comprData)
+	r, err := zlib.NewReader(&b)
+
+	if r != nil {
+		defer r.Close()
+	}
+
+	if err != nil {
+		return err
+	}
+
+	data := make([]byte, uncompressedLength)
+	lenRead := 0
```

Comment:
This temporary buffer allocation is avoidable, like C's `realloc`.

```go
offset := len(cr.bytesBuf)

if cap(cr.bytesBuf) - offset < uncompressedLength {
  old := cr.bytesBuf
  cr.bytesBuf = make([]byte, offset, offset + uncompressedLength)
  copy(cr.bytesBuf, old)
}

data := cr.bytesBuf[offset:offset + uncompressedLength]
```

If you can't understand this, please ignore this comment.  I'll optimize allocation later.